File: python/semanticabi/metadata/EthTransferable.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Optional

from semanticabi.metadata.EthTransferType import EthTransferType


class EthTransferable(ABC):
    """
    Interface for things that can be transferred including root transaction, internal transaction, and token transfers.

    @author zuyezheng
    """

    @property
    @abstractmethod
    def contract_address(self) -> str:
        """ Contract address for the transfer. """
        pass

    @property
    @abstractmethod
    def from_address(self) -> str:
        pass

    @property
    @abstractmethod
    def to_address(self) -> str:
        pass

    @property
    @abstractmethod
    def value(self) -> Optional[int]:
        pass

    @property
    @abstractmethod
    def transfer_type(self) -> EthTransferType:
        pass

File: js/src/semanticabi/metadata/EthTransferable.ts
import {EthTransferType} from "semanticabi/metadata/EthTransferType";
import {Optional} from "common/Optional";

/**
 * Interface for things that can be transferred including root transaction, internal transaction, and token transfers.
 *
 * @author zuyezheng
 */
export interface EthTransferable {

    /**
     * Contract address for the transfer.
     */
    get contractAddress(): string;

    get fromAddress(): string;

    get toAddress(): string;

    get value(): Optional<number>;

    get transferType(): EthTransferType;

}